### DISCUSSION TITLE
How sub-group is arranged inside 2D/3D work group?

### DISCUSSION THREAD

**Discussion by CaoZhongZ** [intel/llvm#12892] [Q&A]
I noticed that sub-group is 1d only construct. So how is it arranged inside a, say, 2d work-group? Is it column major or row major?

**Comment by AlexeySachkov** [comment]
Hi @CaoZhongZ,

The order in which work-items are grouped into sub-groups is implementation defined, see [3.9.4. Work-group data parallel kernels](https://registry.khronos.org/SYCL/specs/sycl-2020/html/sycl-2020.html#_work_group_data_parallel_kernels):

> Work-groups may be further subdivided into [sub-groups](https://registry.khronos.org/SYCL/specs/sycl-2020/html/sycl-2020.html#sub-group). The work-items that compose a sub-group are selected in an implementation-defined way, and therefore the size and number of sub-groups may differ for each kernel. Moreover, different devices may make different guarantees with respect to how sub-groups within a work-group are scheduled. The maximum number of work-items in any sub-group in a kernel is based on a combination of the kernel and its dispatch dimensions. The size of any sub-group in the dispatch is between 1 and this maximum sub-group size, and the size of an individual sub-group is invariant for the duration of a kernel’s execution. Similarly to work-groups, the work-items within the same sub-group can be synchronized by calling the `group_barrier` function.
> 
> Portable device code must not assume that work-items within a sub-group execute in any particular order, that work-groups are subdivided into sub-groups in a specific way, nor that the work-items within a sub-group provide specific forward progress guarantees.

For example, on Intel GPU devices the way of bundling work-items into sub-group may change from kernel to kernel because it is based on some heuristics

**Reply by CaoZhongZ** [reply]
Well, change bundle strategy of subgroup between kernels… Isn't that unnecessarily flexible? Hence discourage user to predict access pattern and hardware behavior. Which means the construct is useless unless someone could assert some control.
